### ISSUE TITLE
Focus restore overridden when using FocusTrap inside Vue Transition

### ISSUE THREAD

**Issue by ragulka** [tailwindlabs/headlessui#2218]
**What package within Headless UI are you using?**

For example: @headlessui/vue

**What version of that package are you using?**

For example: v1.7.7

**What browser are you using?**

For example: N/A

**Reproduction URL**

https://stackblitz.com/edit/vitejs-vite-eszqtg?file=src/App.vue

**Describe your issue**

When `FocusTrap` is inside a Vue `Transition` element, it's possible that restoring focus to the trigger does not work. Focus does get restored, but it is immediately set back to the previously active element inside the FocusTrap by the [`handleBlur` function here](https://github.com/tailwindlabs/headlessui/blob/2f13496acb602479476e321471b051760d75e6e4/packages/%40headlessui-vue/src/components/focus-trap/focus-trap.ts#L124-L137).

This probably happens because the `focusout` event handler (handleBlur) is a native DOM element, which is triggered _after_ the `FocusTrap` component is `unMounted`. Because of the transition, the component gets unmounted, but it's still present in DOM (with the native DOM event handlers) while the transition occurs.

An easy fix for this would be to[ track wether the component is mounted](https://github.com/tailwindlabs/headlessui/blob/2f13496acb602479476e321471b051760d75e6e4/packages/%40headlessui-vue/src/components/focus-trap/focus-trap.ts#L238-L240) and only handle the blur event if the component is still mounted:

```js
let mounted = ref(false)
onMounted(() => (mounted.value = true))
onUnmounted(() => (mounted.value = false))

function handleBlur() {
   if (!mounted.value) return;
   
   // ... handle blur if mounted
}
```

Note: I understand that using `TransitionRoot` may help when using a single FocusTrap, but I could not find a way to make `TransitionRoot` work with **dynamic components** (`<Component :is... />`), which is what my use-case calls for.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your bug report!
Much appreciated! 🙏

Using the `TransitionRoot` helps in different scenario's better because it can communicate with other components from Headless UI.
If you are looking for how to use dynamic components, then you can use the `as` prop:
```js
<TransitionRoot
  :show="currentComponent != null"
  :as="getCurrentComponent(currentComponent)"
  enter-from="opacity-0"
  enter="transition-opacity duration-100"
  enter-to="opacity-100"
  leave-from="opacity-100"
  leave="transition-opacity duration-100"
  leave-to="opacity-0"
/>
```

Here is an updated Stackblitz link: https://stackblitz.com/edit/vitejs-vite-cwodet?file=src/App.vue

Would that work for you?

**Comment by ragulka** [comment]
> If you are looking for how to use dynamic components, then you can use the `as` prop

That's an option I hadn't considered. However, it looks like that won't perform the transition when switching between components - only when the current component is toggled off and on.

Basically, I want to achieve the transitions from one component to another, while ensuring the current component has focus trapped and if I set the component to `null`, focus is restored to what it was before opening any focus-trapped component.

Does that make sense?

**Comment by RobinMalfait** [comment]
I see what you mean now, I could reproduce it and applied a fix for this. It's very similar to what you provided but had to make a few additional changes.

This should be fixed by #2331, and will be available in the next release.

You can already try it using:
- `npm install @headlessui/vue@insiders`.
